### PR TITLE
`defer_measurements` improvements

### PR DESCRIPTION
**Pull request by mudit2812**
**Context:**
`defer_measurements` had some restrictions added to it due to the changes needed for the new features that have been added. This includes not supporting custom wire labels, and some forgotten changes.

**Description of the Change:**
* Added custom qnode transform so that qnode transformations also provide device wires as an argument.
* Added support for custom wire labels. When the object being transformed is a qnode with a device with `wires=None` or a tape, new wires are added to the tape with labels `mv{i}`, where `{i}` is an integer. When a qnode is transformed with a device that does not have `None` wires, users have to specify extra wires for the mid-circuit measurements, and those will be used for storing mid-circuit measurements.
* The transform now updates the wires of `qml.probs`, `qml.sample` and `qml.counts` to the original tape wires if no observable or wires are specified. This ensures that any wires that are added by the transform (or any wires that users add to devices for storing the mid-circuit measurements) do not get included in the measurement results.

TODO:
* [x] Split transform into multiple functions.

**Benefits:**
Custom wire labels are supported and the results for terminal measurements are *more* correct.

**Possible Drawbacks:**
* Users can't use `mv{i}` as wire labels. Though this is not really a drawback since users could not use _any_ custom wire labels before.
* For the changes to measurements, users could get confused when the extra wires they add to the device are not included in the results.
* Moreover, if the extra wires they specify are not the last wires of the device, there will be wires that store mid-circuit measurement results interspersed within the circuit, which can be even more confusing. I've added documentation to make this clear.

**Related GitHub Issues:**

### CODE REVIEW

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -77,6 +77,14 @@
 
 <h3>Improvements 🛠</h3>
 
+* `qml.defer_measurements` now supports custom wire labels. When transforming tapes or QNodes with devices
+  initialized with `wires=None`, wires with labels `mv{i}`, where `{i}` is an integer, will be used to
+  store mid-circuit measurement results. Thus, these labels are reserved when using the `qml.defer_measurements`
+  transform. Additionally, when using measuring `qml.probs`, `qml.sample`, or `qml.counts` without
+  specifying an observable or wires, `qml.defer_measurements` will correctly transform the measurements such
+  that only the original wires are used for the measurements.
```
Comment (trbromley) [inline doc/releases/changelog-dev.md:85]: :+1:

File: doc/releases/changelog-dev.md [hunk 2]
```diff
@@ -77,6 +77,14 @@
 
 <h3>Improvements 🛠</h3>
 
+* `qml.defer_measurements` now supports custom wire labels. When transforming tapes or QNodes with devices
+  initialized with `wires=None`, wires with labels `mv{i}`, where `{i}` is an integer, will be used to
+  store mid-circuit measurement results. Thus, these labels are reserved when using the `qml.defer_measurements`
+  transform. Additionally, when using measuring `qml.probs`, `qml.sample`, or `qml.counts` without
```
Comment (trbromley) [inline doc/releases/changelog-dev.md:83]:
```suggestion
  transform. Additionally, when measuring `qml.probs`, `qml.sample`, or `qml.counts` without
```

File: pennylane/transforms/defer_measurements.py [hunk 3]
```diff
@@ -48,11 +48,46 @@ def defer_measurements(tape: QuantumTape) -> (Sequence[QuantumTape], Callable):
...
 
-        Devices that inherit from :class:`~pennylane.QubitDevice` **must** be initialized
-        with an additional wire for each mid-circuit measurement after which the measured
-        wire is reused or reset for ``defer_measurements`` to transform the quantum tape
-        correctly. Hence, devices and quantum tapes must also be initialized without custom
-        wire labels for correct behaviour.
+        Devices that inherit from :class:`~pennylane.QubitDevice` such as
+        ``default.qubit.legacy`` **must** be initialized with an additional wire for
+        each mid-circuit measurement after which the measured wire is reused or reset
+        for ``defer_measurements`` to transform the quantum tape correctly. Devices
```
Comment (trbromley) [inline pennylane/transforms/defer_measurements.py:54]:
```suggestion
        each mid-circuit measurement that includes a wire reuse or reset. Devices
```

File: pennylane/transforms/defer_measurements.py [hunk 4]
```diff
@@ -48,11 +48,46 @@ def defer_measurements(tape: QuantumTape) -> (Sequence[QuantumTape], Callable):
...
+        that inherit from :class:`~pennylane.devices.Device` such as ``default.qubit``
+        must be initialized with the same number of wires, or with ``wires=None``.
```
Comment (trbromley) [inline pennylane/transforms/defer_measurements.py:56]: Same number as what?

File: pennylane/transforms/defer_measurements.py [hunk 5]
```diff
@@ -48,11 +48,46 @@ def defer_measurements(tape: QuantumTape) -> (Sequence[QuantumTape], Callable):
...
+
+    .. note::
+
+        If using custom wire labels, do not use ``"mv{i}"`` as labels, where ``{i}`` is an
+        integer. These labels are reserved for wires that store mid-circuit measurement results
+        if transforming a tape or executing on a device that has ``wires=None``.
```
Comment (trbromley) [inline pennylane/transforms/defer_measurements.py:62]: Could we rephrase this as:
- If you are specifying wires when loading your device, you'll need to account for the additional wires due to qubit reuse and reset
- If you don't, we'll use `"mv{i}"` for you

File: pennylane/transforms/defer_measurements.py [hunk 6]
```diff
@@ -48,11 +48,46 @@ def defer_measurements(tape: QuantumTape) -> (Sequence[QuantumTape], Callable):
...
+
+            dev = qml.device("default.qubit", wires=4)
+
+            @qml.defer_measurements
+            @qml.qnode(dev)
+            def circuit(x):
+                qml.RX(x, 0)
+                m = qml.measure(0)
+                qml.cond(m, qml.PauliX)(2)
+                return qml.probs()
```
Comment (trbromley) [inline pennylane/transforms/defer_measurements.py:81]: To check, if a user did `wires=[0, 1, 2, 3]`, would this be allowed?

File: pennylane/transforms/defer_measurements.py [hunk 7]
```diff
@@ -48,11 +48,46 @@ def defer_measurements(tape: QuantumTape) -> (Sequence[QuantumTape], Callable):
...
+        that inherit from :class:`~pennylane.devices.Device` such as ``default.qubit``
+        must be initialized with the same number of wires, or with ``wires=None``.
+
+    .. note::
+
+        If using custom wire labels, do not use ``"mv{i}"`` as labels, where ``{i}`` is an
+        integer. These labels are reserved for wires that store mid-circuit measurement results
+        if transforming a tape or executing on a device that has ``wires=None``.
+
+    .. note::
```
Comment (trbromley) [inline pennylane/transforms/defer_measurements.py:64]: This note feels better suited to usage details. Notes should be pretty short.

**Comment by trbromley** [comment]
@mudit2812 do we need this for the release?

**Comment by mudit2812** [comment]
> @mudit2812 do we need this for the release?

No, but it would be nice to have.

**Comment by trbromley** [comment]
> > @mudit2812 do we need this for the release?
> 
> No, but it would be nice to have.

Looks worthwhile if we can do it. Let us know if you need help from the rest of the team

File: pennylane/transforms/defer_measurements.py [hunk 8]
```diff
@@ -192,10 +244,18 @@ def func(x, y):
...
+    # Map any mid-circuit measurements to the correct wires that store them
     for mp in tape.measurements:
         if mp.mv is not None:
             wire_map = {m.wires[0]: control_wires[m.id] for m in mp.mv.measurements}
             mp = qml.map_wires(mp, wire_map=wire_map)
+
+        elif isinstance(
+            mp,
+            (qml.measurements.SampleMP, qml.measurements.ProbabilityMP, qml.measurements.CountsMP),
+        ) and not (mp.obs or mp.wires):
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/transforms/defer_measurements.py:287]: The presence of `mp.obs` should show up in `mp.wires`. 

```
>>> qml.sample(op=qml.PauliX(0)).wires
<Wires = [0]>
```

File: pennylane/transforms/defer_measurements.py [hunk 9]
```diff
@@ -26,6 +26,76 @@
 # pylint: disable=too-many-branches, too-many-statements
```
Comment (albi3ro) [inline pennylane/transforms/defer_measurements.py:26]: Do we still need these pylint silencings?

Comment (mudit2812) [reply pennylane/transforms/defer_measurements.py:26]: Not anymore. Forgot to remove them 😄

File: pennylane/transforms/defer_measurements.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (mudit2812) [reply pennylane/transforms/defer_measurements.py:321]: Not anymore, because I'm just unconditionally raising an error prior to this code block if `qml.state` is present.

**Comment by trbromley** [comment]
Thanks @mudit2812! Closing this for now, but let's make sure to record these improvements in Shortcut so that we can draw from the backlog in future.
